Adopt smart pointers in ServiceWorkerDownloadTask, ServiceWorkerFetchTask, ServiceWorkerNavigationPreloader

jelee53 opened the pull request:
#### bd028ed247b4e93e0d8f3ee8048d459eddd4a49b
<pre>
Adopt smart pointers in ServiceWorkerDownloadTask, ServiceWorkerFetchTask, ServiceWorkerNavigationPreloader
<a href="https://bugs.webkit.org/show_bug.cgi?id=295445">https://bugs.webkit.org/show_bug.cgi?id=295445</a>
<a href="https://rdar.apple.com/154996571">rdar://154996571</a>

Reviewed by Chris Dumez and Per Arne Vollan.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

* Source/WebKit/NetworkProcess/NetworkSocketChannel.cpp:
(WebKit::NetworkSocketChannel::NetworkSocketChannel):
(WebKit::NetworkSocketChannel::~NetworkSocketChannel):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::start):
(WebKit::ServiceWorkerDownloadTask::didReceiveData):
(WebKit::ServiceWorkerDownloadTask::didFinish):
(WebKit::ServiceWorkerDownloadTask::didFailDownload):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::ServiceWorkerFetchTask):
(WebKit::ServiceWorkerFetchTask::loadResponseFromPreloader):
(WebKit::ServiceWorkerFetchTask::checkedPreloader):
(WebKit::ServiceWorkerFetchTask::loadBodyFromPreloader):
(WebKit::ServiceWorkerFetchTask::cancelPreloadIfNecessary):
(WebKit::ServiceWorkerFetchTask::convertToDownload):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerNavigationPreloader.cpp:
(WebKit::ServiceWorkerNavigationPreloader::loadFromNetwork):
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/297086@main">https://commits.webkit.org/297086@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2835962198343fe1306bbf1044c42f42859a1135

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110510 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30169 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116536 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30848 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38757 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84033 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113458 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24624 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99506 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64474 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23986 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60331 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94005 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17705 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119326 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37550 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27867 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92999 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37924 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95777 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92822 "Found 3 new API test failures: /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy, /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed, /WebKitGTK/TestResources:/webkit/WebKitWebResource/get-data-error (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37828 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15568 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17828 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37445 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42916 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37107 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40447 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38816 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkSocketChannel.cpp:
```diff
@@ -57,13 +57,15 @@ NetworkSocketChannel::NetworkSocketChannel(NetworkConnectionToWebProcess& connec
     relaxAdoptionRequirement();
-    if (!m_session)
+
+    CheckedPtr checkedSession = m_session.get();
```
`checkedSession` -> `session`

jelee53 replied:
Hi Chris - there is already a parameter in the function called `session`. 

Renaming results in this error:
```
In file included from /Users/jessicalee/Documents/webkit-col/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/UnifiedSource41.cpp:6:
/Users/jessicalee/Documents/webkit-col/OpenSource/Source/WebKit/NetworkProcess/NetworkSocketChannel.cpp:61:16: error: redefinition of 'session'
   61 |     CheckedPtr session = m_session.get();
```

Would you prefer that I rename the function parameter `session`  -> `networkSession` so that the CheckedPtr can be called `session`? 

e.g 
```
+++ b/Source/WebKit/NetworkProcess/NetworkSocketChannel.cpp
@@ -49,20 +49,20 @@ RefPtr<NetworkSocketChannel> NetworkSocketChannel::create(NetworkConnectionToWeb
     return result;
 }
 
-NetworkSocketChannel::NetworkSocketChannel(NetworkConnectionToWebProcess& connection, NetworkSession* session, const ResourceRequest& request, const String& protocol, WebSocketIdentifier identifier, WebPageProxyIdentifier webPageProxyID, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, const WebCore::ClientOrigin& clientOrigin, bool hadMainFrameMainResourcePrivateRelayed, bool allowPrivacyProxy, OptionSet<AdvancedPrivacyProtections> advancedPrivacyProtections, WebCore::StoredCredentialsPolicy storedCredentialsPolicy)
+NetworkSocketChannel::NetworkSocketChannel(NetworkConnectionToWebProcess& connection, NetworkSession* networkSession, const ResourceRequest& request, const String& protocol, WebSocketIdentifier identifier, WebPageProxyIdentifier webPageProxyID, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, const WebCore::ClientOrigin& clientOrigin, bool hadMainFrameMainResourcePrivateRelayed, bool allowPrivacyProxy, OptionSet<AdvancedPrivacyProtections> advancedPrivacyProtections, WebCore::StoredCredentialsPolicy storedCredentialsPolicy)
     : m_connectionToWebProcess(connection)
     , m_identifier(identifier)
-    , m_session(session)
+    , m_session(networkSession)
     , m_errorTimer(*this, &NetworkSocketChannel::sendDelayedError)
     , m_webPageProxyID(webPageProxyID)
 {
     relaxAdoptionRequirement();
 
-    CheckedPtr checkedSession = m_session.get();
+    CheckedPtr session = m_session.get();
     if (!session)
         return;
 
-    m_socket = checkedSession->createWebSocketTask(webPageProxyID, frameID, pageID, *this, request, protocol, clientOrigin, hadMainFrameMainResourcePrivateRelayed, allowPrivacyProxy, advancedPrivacyProtections, storedCredentialsPolicy);
+    m_socket = session->createWebSocketTask(webPageProxyID, frameID, pageID, *this, request, protocol, clientOrigin, hadMainFrameMainResourcePrivateRelayed, allowPrivacyProxy, advancedPrivacyProtections, storedCredentialsPolicy);
     if (CheckedPtr socket = m_socket.get()) {
 #if PLATFORM(COCOA)
         session->addWebSocketTask(webPageProxyID, *socket);
```

cdumez replied:
If there is already a parameter for `session` then use it in this function, you do not need a local smart pointer for parameters.

cdumez commented on Source/WebKit/NetworkProcess/NetworkSocketChannel.cpp:
```diff
@@ -74,7 +74,7 @@ NetworkSocketChannel::~NetworkSocketChannel()
 #if PLATFORM(COCOA)
         if (RefPtr sessionSet = m_session ? socket->sessionSet() : nullptr)
-            m_session->removeWebSocketTask(*sessionSet, *socket);
+            CheckedPtr { m_session.get() }->removeWebSocketTask(*sessionSet, *socket);
```
Since we know that `m_session` is non-null, it may be more efficient / clearer to use:
```
CheckedRef { *m_session }->removeWebSocketTask(*sessionSet, *socket);
```

cdumez commented on Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
```diff
@@ -178,7 +178,7 @@ void ServiceWorkerDownloadTask::start()
 
     CheckedRef manager = m_networkProcess->downloadManager();
-    Ref download = Download::create(manager.get(), m_downloadID, *this, *networkSession());
+    Ref download = Download::create(manager.get(), m_downloadID, *this, CheckedRef { *networkSession() }.get() );
```
do we really need the `.get()`?

jelee53 replied:
We don't. I was trying to clean this up in a follow-up commit but I was getting this particular error when I tried to fix it.
```
   70 |     FOR_EACH_WEBKIT_PREFERENCE(DECLARE_PREFERENCE_GETTER_AND_SETTERS)
/Users/jessicalee/Documents/webkit-col/OpenSource/WebKitBuild/Debug/DerivedSources/WebKit/WebPreferencesFeatures.cpp:3721:34: error: reference to type 'const bool' could not bind to an rvalue of type 'void'
```

I'm doing a `make clean` and a fresh `make debug` on main to see if that will get rid of the issue. But yes, I will fix this.

cdumez reviewed (approved):
LGTM with nits.

jelee53 committed:
Adopt smart pointers in ServiceWorkerDownloadTask, ServiceWorkerFetchTask, ServiceWorkerNavigationPreloader
https://bugs.webkit.org/show_bug.cgi?id=295445
rdar://154996571

Reviewed by Chris Dumez and Per Arne Vollan.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

* Source/WebKit/NetworkProcess/NetworkSocketChannel.cpp:
(WebKit::NetworkSocketChannel::NetworkSocketChannel):
(WebKit::NetworkSocketChannel::~NetworkSocketChannel):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::start):
(WebKit::ServiceWorkerDownloadTask::didReceiveData):
(WebKit::ServiceWorkerDownloadTask::didFinish):
(WebKit::ServiceWorkerDownloadTask::didFailDownload):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::ServiceWorkerFetchTask):
(WebKit::ServiceWorkerFetchTask::loadResponseFromPreloader):
(WebKit::ServiceWorkerFetchTask::checkedPreloader):
(WebKit::ServiceWorkerFetchTask::loadBodyFromPreloader):
(WebKit::ServiceWorkerFetchTask::cancelPreloadIfNecessary):
(WebKit::ServiceWorkerFetchTask::convertToDownload):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerNavigationPreloader.cpp:
(WebKit::ServiceWorkerNavigationPreloader::loadFromNetwork):
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:

Canonical link: https://commits.webkit.org/297086@main
